Modernize cpp

pb-dseifert opened the pull request:
* Makefile should adhere to common Linux packaging conventions
* Allow using all APIs without calling `strlen()` internally
* Modernize to C++ code:
  * Avoid manual memory management with `new`/`delete` and use RAII types instead
  * Avoid passing pointers when references suffice
  * Avoid references for input values
  * Reduce code duplication in the `Align()` function

pb-dseifert commented on src/ssw_cpp.cpp:
```diff
@@ -5,29 +5,25 @@
-  //   A     C            G
-    4, 0, 4, 1,  4, 4, 4, 2,  4, 4, 4, 4,  4, 4, 4, 4,
-  //             T
-    4, 4, 4, 4,  3, 0, 4, 4,  4, 4, 4, 4,  4, 4, 4, 4,
```
@mengyao I assume the `0` after the `t`/`T`s are a typo?

mengyao replied:
yes, these are typos. thanks!

mengyao commented on src/ssw_cpp.cpp:
```diff
@@ -440,32 +398,24 @@ bool Aligner::ReBuild(
 bool Aligner::ReBuild(
     const int8_t* score_matrix,
-    const int&    score_matrix_size,
+    const int /*score_matrix_size*/,
```
Why this parameter is commented off, and we changed to use score_matrix_size_ later?

pb-dseifert replied:
I've changed it to assign the value and use that now

pb-dseifert committed:
Modernize code

* All APIs should be callable with size parameters.

pb-dseifert committed:
Fix Makefile for modern packaging conventions

pb-dseifert committed:
to_cigar_int(): use unsigned char argument

* `char` should always be converted to `unsigned char` before
  converting to larger sizes:
  https://wiki.sei.cmu.edu/confluence/display/c/STR34-C.+Cast+characters+to+unsigned+char+before+converting+to+larger+integer+sizes

mengyao commented:
Merged your PR! thanks!
